src/cluster.h

```diff
@@ -173,257 +51,102 @@ typedef struct {
-static_assert(offsetof(clusterMsg, type) == 12, "unexpected field offset");
-static_assert(offsetof(clusterMsg, count) == 14, "unexpected field offset");
-static_assert(offsetof(clusterMsg, currentEpoch) == 16, "unexpected field offset");
-static_assert(offsetof(clusterMsg, configEpoch) == 24, "unexpected field offset");
-static_assert(offsetof(clusterMsg, offset) == 32, "unexpected field offset");
-static_assert(offsetof(clusterMsg, sender) == 40, "unexpected field offset");
-static_assert(offsetof(clusterMsg, myslots) == 80, "unexpected field offset");
-static_assert(offsetof(clusterMsg, slaveof) == 2128, "unexpected field offset");
-static_assert(offsetof(clusterMsg, myip) == 2168, "unexpected field offset");
-static_assert(offsetof(clusterMsg, extensions) == 2214, "unexpected field offset");
-static_assert(offsetof(clusterMsg, notused1) == 2216, "unexpected field offset");
-static_assert(offsetof(clusterMsg, pport) == 2246, "unexpected field offset");
-static_assert(offsetof(clusterMsg, cport) == 2248, "unexpected field offset");
-static_assert(offsetof(clusterMsg, flags) == 2250, "unexpected field offset");
-static_assert(offsetof(clusterMsg, state) == 2252, "unexpected field offset");
-static_assert(offsetof(clusterMsg, mflags) == 2253, "unexpected field offset");
-static_assert(offsetof(clusterMsg, data) == 2256, "unexpected field offset");
-
-#define CLUSTERMSG_MIN_LEN (sizeof(clusterMsg)-sizeof(union clusterMsgData))
-
-/* Message flags better specify the packet content or are used to
- * provide some information about the node state. */
-#define CLUSTERMSG_FLAG0_PAUSED (1<<0) /* Master paused for manual failover. */
-#define CLUSTERMSG_FLAG0_FORCEACK (1<<1) /* Give ACK to AUTH_REQUEST even if
-                                            master is up. */
-#define CLUSTERMSG_FLAG0_EXT_DATA (1<<2) /* Message contains extension data */
-
 /* ---------------------- API exported outside cluster.c -------------------- */
+/*
+ * Functions requiring per-clustering mechanism implementation: Lifecycle events
+ */
 void clusterInit(void);
 void clusterInitListeners(void);
 void clusterCron(void);
 void clusterBeforeSleep(void);
-clusterNode *getNodeByQuery(client *c, struct redisCommand *cmd, robj **argv, int argc, int *hashslot, int *ask);
-int verifyClusterNodeId(const char *name, int length);
-clusterNode *clusterLookupNode(const char *name, int length);
-int clusterRedirectBlockedClientIfNeeded(client *c);
-void clusterRedirectClient(client *c, clusterNode *n, int hashslot, int error_code);
-void migrateCloseTimedoutSockets(void);
 int verifyClusterConfigWithData(void);
+void clusterUpdateMyselfFlags(void);
 unsigned long getClusterConnectionsCount(void);
+void clusterUpdateMyselfHostname(void);
+void clusterUpdateMyselfAnnouncedPorts(void);
+void clusterUpdateMyselfIp(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: cluster commands
+ */
+void clusterCommand(client *c);
```

Comment:
I would expect we would split this into the application APIs (readonly, readwrite, slots, shards) and then the application APIs.

Reply:
Yeah, I also would like to do that. Of the various cluster sub-commands...
* some make no sense in flotilla
* I genericized the cluster slots implementation and moved into cluster.c
* There are a few more that have generic implementations that can be shared
* cluster shards is a challenge. It's implemented in a manner that makes it very hard to abstract and make shared.

Here is a cut-and-paste from some notes I took ("return error" means it is N/A in flotilla:

"ADDSLOTS <slot> [<slot> ...]", - RETURN ERROR
"ADDSLOTSRANGE <start slot> <end slot> [<start slot> <end slot> ...]", RETURN ERROR
"BUMPEPOCH", RETURN ERROR
"COUNT-FAILURE-REPORTS <node-id>", RETURN ERROR
"COUNTKEYSINSLOT <slot>", SHARED IMPLEMENT
"DELSLOTS <slot> [<slot> ...]", RETURN ERROR
"DELSLOTSRANGE <start slot> <end slot> [<start slot> <end slot> ...]", RETURN ERROR
"FAILOVER [FORCE|TAKEOVER]", ? SHORT TERM NO FORCE, LONG TERM - YES
"FORGET <node-id>", RETURN ERROR
"GETKEYSINSLOT <slot> <count>", SHARED IMPL
"FLUSHSLOTS", RETURN ERROR
"INFO", IMPLEMENT
"KEYSLOT <key>", SHARED CODE
"MEET <ip> <port> [<bus-port>]", RETURN ERROR
"MYID", IMPLEMENT
"MYSHARDID", IMPLEMENT
"NODES", IMPLEMENT BUT MOST OF THE OUTPUT IS N/A
"    Return cluster configuration seen by node. Output format:",
"    <id> <ip:port@bus-port[,hostname]> <flags> <master> <pings> <pongs> <epoch> <link> <slot> ...",
"REPLICATE <node-id>", RETURN ERROR
"RESET [HARD|SOFT]", RETURN ERROR
"SET-CONFIG-EPOCH <epoch>", RETURN ERROR
"SETSLOT <slot> (IMPORTING <node-id>|MIGRATING <node-id>|STABLE|NODE <node-id>)", RETURN ERROR
"REPLICAS <node-id>", IMPLEMENT
"SAVECONFIG", MAYBE EVENTUALLY
"SLOTS", IMPLEMENT
"SHARDS", IMPLEMENT
"LINKS", RETURN ERROR

In addition we will probably want to add commands.

In short, I totally agree with this comment and I am happy to work on doing this but the PR up to this point represents my attempt to get something ready that we can experiment with. I can do this now, in this PR, or later. What's your take?

Reply:
I would like to work on this now I think. What do you mean by readonly and readwrite?

Reply:
> What do you mean by readonly and readwrite?

My brain was just thinking about the client APIs for the cluster spec, of which `READWRITE` and `READONLY` are there but have nothing to do with the cluster command. 

> "SAVECONFIG", MAYBE EVENTUALLY

I don't know why this would be shared, flotilla shouldn't care?

> FAILOVER

Maybe we can simplify this and just return OK for the arguments we silently ignore. For example AWS used the force argument forever because we didn't really think it through. 

> NODES

We should make a decision if we want to maintain this or not. It's supposed to be an admin command, I'm fine not having it be supported in flotilla and keep pushing clients to adopt slots or shards.